server/service.go

```diff
@@ -0,0 +1,72 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/benchttp/engine/internal/socketio"
+	"github.com/benchttp/engine/runner"
+)
+
+type service struct {
+	mu     sync.RWMutex
+	runner *runner.Runner
+	cancel context.CancelFunc
+}
+
+// doRun calls runner.Runner.Run. The service state is overwritten.
+// The return value of runner.Runner.Run is send to the client via
+// w. The run progress is streamed through w.
+func (s *service) doRun(w socketio.Writer, cfg runner.Config) {
+	ctx, cancel := context.WithCancel(context.Background())
+	s.cancel = cancel
+
+	s.runner = runner.New(s.sendRecordingProgess(w))
+
+	out, err := s.runner.Run(ctx, cfg)
+	if err != nil {
+		_ = w.WriteJSON(errorMessage{Event: "done", Error: err.Error()})
+		return
+	}
+
+	_ = w.WriteJSON(doneMessage{Event: "done", Data: *out})
+}
+
+// cancelRun cancels the run of the current runner.
+// If the runner is nil, cancelRun is noop.
+func (s *service) cancelRun() (ok bool) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.runner == nil {
+		return false
+	}
+	s.cancel()
```

Comment:
question: this panics if `s.cancel == nil`, is it the behaviour we want?

Reply:
From my manual testing, `service.flush` is never called on empty state.

I didn't have time to write tests for it yet.

I will investigate. And wrap it in a conditional call for the mergeablility of the PR.

Reply:
As I understand, having a non-nil runner with a nil `cancel` would be the result of a design flaw. In that sense letting it panic could be an option I guess 🤷